torch/distributed/algorithms/join.py

```diff
@@ -0,0 +1,215 @@
+import warnings
+from abc import ABC, abstractmethod
+from typing import List
+
+import torch
+import torch.distributed as dist
+
+
+class _JoinHook(ABC):
+    r"""
+    This defines a join hook, which provides two entry points in the join
+    context manager: a main hook, which is called repeatedly while there exists
+    a non-joined process, and a post-hook, which is called once all processes
+    have joined.
+
+    To implement a join hook for the generic join context manager, define a
+    class that inherits from :class:`_JoinHook`, override ``main_hook()`` and
+    ``post_hook()`` as appropriate, and override ``device()`` and
+    ``process_group()`` to provide the device and process group information,
+    respectively, which are needed for the join context manager implementation.
+    """
+    def main_hook(self):
```

Comment:
Curious, any reason these are not abstract method? Is it because it's possible that subclass does not need to override it?

Reply:
Yup, that was the idea. I was not sure if it was better to make these abstract and have the child class explicitly override with a vacuous implementation or to just leave these as non-abstract. Perhaps, it could be better to make these abstract as well so that the implementer is intentionally leaving them as vacuous if they are not being used. I think this comes down to personal preference.